File: go/deliberation/pkg/deliberation/fixtures_test.go
// Cross-language fixture parity tests.
//
// Reads the SAME JSON fixture files as the Node @accretive-substrate/deliberation test
// suite (vendored copies at internal/test/fixtures/) and asserts the
// same expected_voices / expected_adjudication / expected_final_verdict.
//
// Identical assertions in Node + Go prove the cross-pipeline parity claim
// for Path B. If a future B2.Go specialist diverges from B2.Node, the
// fixture tests on both sides fail simultaneously.
package deliberation_test

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/ramene/accretive-substrate/go/deliberation/pkg/deliberation"
	"github.com/ramene/accretive-substrate/go/deliberation/pkg/deliberation/specialists"
)

type Fixture struct {
	Name                 string                       `json:"name"`
	Description          string                       `json:"description"`
	EvidencePackage      deliberation.EvidencePackage `json:"evidence_package"`
	ExpectedVoices       map[string]string            `json:"expected_voices,omitempty"`
	ExpectedAdjudication *struct {
		Verdict string `json:"verdict"`
	} `json:"expected_adjudication,omitempty"`
	ExpectedGuardrail *struct {
		PolicyApplied string `json:"policy_applied"`
	} `json:"expected_guardrail,omitempty"`
	ExpectedFinalVerdict string `json:"expected_final_verdict,omitempty"`
}

func loadFixtures(t *testing.T) []Fixture {
	t.Helper()
	dir := filepath.Join("..", "..", "internal", "test", "fixtures")
	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Fatalf("read fixture dir: %v", err)
	}
	var out []Fixture
	for _, e := range entries {
		if !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		raw, err := os.ReadFile(filepath.Join(dir, e.Name()))
		if err != nil {
			t.Fatalf("read %s: %v", e.Name(), err)
		}
		var f Fixture
		if err := json.Unmarshal(raw, &f); err != nil {
			t.Fatalf("unmarshal %s: %v", e.Name(), err)
		}
		out = append(out, f)
	}
	return out
}

func TestFixtureParity(t *testing.T) {
	// B2.Go: all 8 specialists present. Cross-language parity now asserts
	// EVERY voice + full adjudication + guardrail + final verdict against
	// the same fixture files as the Node @accretive-substrate/deliberation suite.
	// This is the cross-language parity proof for Path B.

	for _, f := range loadFixtures(t) {
		t.Run(f.Name, func(t *testing.T) {
			result := deliberation.RunDeliberation(context.Background(),
				deliberation.RunInput{Evidence: f.EvidencePackage, SourcePipeline: deliberation.PipelineGo},
				deliberation.RunOptions{Specialists: specialists.Active},
			)

			// Verify every voice named in expected_voices matches.
			for name, expected := range f.ExpectedVoices {
				var got string
				for _, v := range result.Voices {
					if v.Specialist == name {
						got = string(v.Verdict)
					}
				}
				if got == "" {
					t.Errorf("specialist %s did not vote", name)
					continue
				}
				if got != expected {
					t.Errorf("%s: expected %s, got %s", name, expected, got)
				}
			}

			// Adjudication.
			if f.ExpectedAdjudication != nil && f.ExpectedAdjudication.Verdict != "" {
				if string(result.Adjudication.Verdict) != f.ExpectedAdjudication.Verdict {
					t.Errorf("adjudication: expected %s, got %s",
						f.ExpectedAdjudication.Verdict, result.Adjudication.Verdict)
				}
			}

			// Guardrail.
			if f.ExpectedGuardrail != nil && f.ExpectedGuardrail.PolicyApplied != "" {
				if result.Guardrail.PolicyApplied != f.ExpectedGuardrail.PolicyApplied {
					t.Errorf("guardrail policy: expected %s, got %s",
						f.ExpectedGuardrail.PolicyApplied, result.Guardrail.PolicyApplied)
				}
			}

			// Final verdict.
			if f.ExpectedFinalVerdict != "" {
				if string(result.FinalVerdict) != f.ExpectedFinalVerdict {
					t.Errorf("final_verdict: expected %s, got %s",
						f.ExpectedFinalVerdict, result.FinalVerdict)
				}
			}
		})
	}
}

File: node/deliberation/test/fixtures.test.mjs
/**
 * Fixture-based integration tests. Each JSON file in test/fixtures/ has an
 * evidence_package + expected_voices + expected_adjudication + expected_final_verdict.
 *
 * The SAME fixture files are loaded by the Go test suite at
 * <this repo>/go/deliberation/internal/test/fixtures/ (vendored copies, kept
 * in sync by B6 tuning loop). This proves cross-language parity.
 */
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync, readdirSync } from 'node:fs';
import { join } from 'node:path';
import { runDeliberation } from '../src/orchestrator.mjs';

const FIXTURE_DIR = new URL('./fixtures/', import.meta.url).pathname;

function loadFixtures() {
  return readdirSync(FIXTURE_DIR)
    .filter(f => f.endsWith('.json'))
    .map(f => ({
      filename: f,
      ...JSON.parse(readFileSync(join(FIXTURE_DIR, f), 'utf8')),
    }));
}

for (const fixture of loadFixtures()) {
  test(`fixture: ${fixture.name} (${fixture.filename})`, async () => {
    const r = await runDeliberation(
      { evidence: fixture.evidence_package, source_pipeline: 'node' },
      { persist: async () => ({ id: 1, ts: new Date(), ok: true }) },
    );

    // Voice expectations (subset — only check specialists named in fixture).
    if (fixture.expected_voices) {
      for (const [name, expectedVerdict] of Object.entries(fixture.expected_voices)) {
        const voice = r.voices.find(v => v.specialist === name);
        assert.ok(voice, `missing voice for specialist ${name}`);
        assert.equal(
          voice.verdict,
          expectedVerdict,
          `${fixture.name}: ${name} expected ${expectedVerdict}, got ${voice.verdict} (${voice.rationale})`,
        );
      }
    }

    // Adjudication.
    if (fixture.expected_adjudication?.verdict) {
      assert.equal(
        r.adjudication.verdict,
        fixture.expected_adjudication.verdict,
        `${fixture.name}: adjudication expected ${fixture.expected_adjudication.verdict}, got ${r.adjudication.verdict}`,
      );
    }

    // Guardrail.
    if (fixture.expected_guardrail?.policy_applied) {
      assert.equal(
        r.guardrail.policy_applied,
        fixture.expected_guardrail.policy_applied,
        `${fixture.name}: guardrail policy expected ${fixture.expected_guardrail.policy_applied}, got ${r.guardrail.policy_applied}`,
      );
    }

    // Final verdict.
    if (fixture.expected_final_verdict) {
      assert.equal(
        r.final_verdict,
        fixture.expected_final_verdict,
        `${fixture.name}: final expected ${fixture.expected_final_verdict}, got ${r.final_verdict}`,
      );
    }
  });
}
